crates/core_arch/src/wasm32/simd128.rs

```diff
@@ -217,6 +325,103 @@ pub const fn v128_const(
     }
 }
 
+/// Returns a new vector with lanes selected from the lanes of the two input
+/// vectors `$a` and `$b` specified in the 16 immediate operands.
+///
+/// The `$a` and `$b` expressions must have type `v128`, and this macro
+/// generates a wasm instruction that is encoded with 16 bytes providing the
+/// indices of the elements to return. The indices `i` in range [0, 15] select
+/// the `i`-th element of `a`. The indices in range [16, 31] select the `i -
+/// 16`-th element of `b`.
+///
+/// Note that this is a macro due to the codegen requirements of all of the
+/// index expressions `$i*` must be constant. A compiler error will be
+/// generated if any of the expressions are not constant.
+///
+/// All indexes `$i*` must have the type `u32`.
+#[allow_internal_unstable(platform_intrinsics, rustc_attrs)]
+pub macro v8x16_shuffle(
```

Comment:
It's probably worth expanding a bit on what I'm thinking here as well. My hope with this is that we'll enable a platform-specific macro which is *not* available in the prelude bit instead must be imported:

```rust
use std::arch::wasm32::v8x16_shuffle;
```

This will have the macro clearly scoped to wasm32, and using a macro (as can be seen below) allows us to still control the implementation in such a way that's opaque to the caller, but also in a way where we can satisfy the compiler's codegen requirements.  Namely all the immediate indices here can be required as constant by the compiler itself, ensuring that we can actually codegen the instruction to LLVM and other backends and such.

I believe that we have all the pieces necessary to have unstable internals that aren't actually exposed to the user. We can also get function-like type checking by carefully crafting the macro to ensure that the input expressions all have an expected type.

Like I mentioned in the PR description though this is, AFAIK, a new thing to do. I don't believe we've pioneered this elsewhere so I don't know if there are hidden downsides (such as using `macro` instead of `macro_rules!`). I'm pretty certain we're not altering the current stability story of libcore, and I think we can continue to figure out the best way to expose an intrinsic like this over time.

FWIW the clang implementation uses a macro that calls a `__builtin_*` function, as does many x86_64 intrinsics with constant arguments where they're all required to be macros in C.

Reply:
I believe that the proper way to expose this kind of functionality in the long term is to use const generics. I hope that we will eventually be able to use `where` clauses to constrain the const values to be within a certain range, which is needed to avoid LLVM-level ICEs due to bad intrinsic inputs.

Reply:
Since this is all unstable I'm happy to merge it as it is, but this is something that we should definitely revisit prior to stabilization.

Reply:
Oh nice! I hadn't thought about the use of const generics here (haven't had a chance to use them myself yet!)

I tested this out and it works well for ` v8x16_shuffle` implementation. It turns out that `simd_shuffle16` in the compiler actually protects us against out-of-bounds indices (they're compile time errors), so I think that happens either with the macro or with the const-generic function. (although this would be best with `where` clauses rather than relying on the just-happens-to-be-caught-during-codegen behavior)

One thing I had difficult though was implementing a const-generic `v16x8_shuffle` in terms of `v8x16_shuffle`. I'm sure it's possible to implement natively but this would mean that we probably would want to include `v16x8_shuffle` as an "intrinsic" since it wouldn't be easily implementable externally.

Reply:
Actually, I've gone ahead and pushed up a version that uses const generics instead. I went ahead and propagated this to everything else using `#[rustc_args_required_const]` in the simd module as well. Given the trajectory of const generics it looks like this may work well for wasm simd!